lib/hyacinth/digital_object/resource_attributes.rb

```diff
@@ -11,51 +11,67 @@ def resource_attributes
         self.class.send(__method__) # class method name same as instance method
       end
 
+      def resource_attribute_names
+        self.class.send(__method__) # class method name same as instance method
+      end
+
       def resource_import_attributes
         self.class.send(__method__) # class method name same as instance method
       end
 
       def resources
-        @resources ||= resource_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @resources ||= resource_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
       end
 
       def resource_imports
-        @resource_imports ||= resource_import_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @resource_imports ||= resource_import_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
       end
 
       def old_resources
-        @old_resources ||= resource_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @old_resources ||= resource_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
       end
 
       def deleted_resources
-        @deleted_resources ||= resource_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @deleted_resources ||= resource_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
+      end
+
+      def delete_resource(resource_name)
+        return unless self.resources[resource_name].present?
+
+        self.deleted_resources[resource_name] = self.resources[resource_name]
+        self.resources[resource_name] = nil
       end
 
       module ClassMethods
         def resource_attributes
-          @resource_attributes ||= Set.new # initialized here because it may not have been initialized in a subclass of the including class
+          @resource_attributes ||= {} # initialized here because it may not have been initialized in a subclass of the including class
           if self.superclass.respond_to?(:resource_attributes)
             @resource_attributes.merge(self.superclass.resource_attributes)
           else
             @resource_attributes
           end
         end
 
+        def resource_attribute_names
+          resource_attributes.keys.dup.freeze
+        end
+
         def resource_import_attributes
-          @resource_import_attributes ||= Set.new # initialized here because it may not have been initialized in a subclass of the including class
+          @resource_import_attributes ||= {} # initialized here because it may not have been initialized in a subclass of the including class
```

Comment:
For later: If we don't end up needing config for resource_import_attributes, this should go back to `Set.new`.